api/ratelimit/v1alpha1/ratelimit_types.go

```diff
@@ -0,0 +1,149 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// Ratelimit allows the user to limit the number of incoming requests
+// to a predefined value based on attributes within the traffic flow.
+type Ratelimit struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RatelimitSpec `json:"spec"`
+
+	// Status defines the current state of Ratelimit.
+	Status RatelimitStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// RatelimitList contains a list of Ratelimit resources.
+type RatelimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Ratelimit `json:"items"`
+}
+
+// RatelimitSpec defines the desired state of Ratelimit
+type RatelimitSpec struct {
+	// Type decides the scope for the Ratelimits.
+	Type RatelimitType `json:"type"`
+	// Rules are a list of Ratelimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RatelimitRule `json:"rules"`
```

Comment:
From reviewing the Envoy [local](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter) and [global](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ratelimit/v3/rls.proto#envoy-v3-api-file-envoy-service-ratelimit-v3-rls-proto) rate limiting docs, I don't understand how `Rules` can be shared across both RL types. For example, [token_bucket](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket) is configured for local rate limiting while a [descriptor](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/common/ratelimit/v3/ratelimit.proto#envoy-v3-api-msg-extensions-common-ratelimit-v3-ratelimitdescriptor) is used for global rate limiting.

Reply:
Rules are not shared, `Type` is a top level field